src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -87,10 +90,29 @@ public bool ReverseOrder
         }
     }
 
+    // The encoding and decoding of LDAP directory controls interprets BER INTEGER values as 32-bit signed integers.
+    // Although BER INTEGERS can exceed this data type's maximum value, previous versions of DirectoryControl (and
+    // its derived classes) encoded and decoded these types by passing the "i" format specifier to BerConverter. The
+    // .NET Framework continues to do so. There is therefore historical precedent to justify limiting BER INTEGER
+    // values to this data type when using AsnDecoder and AsnWriter.
     public class DirectoryControl
     {
+        // Scratch buffer allocations with sizes which are below this threshold should be made on the stack.
+        // This is based on the Active Directory schema. The largest attribute name here is msDS-FailedInteractiveLogonCountAtLastSuccessfulLogon,
+        // which is 53 characters long. This is rounded up to the nearest power of two.
+        internal const int AttributeNameStackAllocationThreshold = 64;
+
+        internal static readonly UTF8Encoding s_utf8Encoding = new(false, true);
+
         internal byte[] _directoryControlValue;
 
+        [ThreadStatic]
+        private static AsnWriter? t_writer;
+
+        [MemberNotNull(nameof(t_writer))]
+        internal static AsnWriter GetWriter()
+            => t_writer ??= new AsnWriter(AsnEncodingRules.BER);
```

Comment:
The correct behavior for every caller is to call `Reset()` on the writer when they get it, because they don't know if they have one that was abandoned due to an exception.

Maybe `GetWriter` should do that for them, instead?

Reply:
Thanks, I've shifted this around.